src/java.base/share/classes/sun/security/provider/SHA3.java

```diff
@@ -91,40 +95,56 @@ void implCompress(byte[] b, int ofs) {
 
     @IntrinsicCandidate
     private void implCompress0(byte[] b, int ofs) {
-       for (int i = 0; i < buffer.length; i++) {
-           state[i] ^= b[ofs++];
-       }
-       keccak();
+        for (int i = 0; i < blockSize / 8; i++) {
+            state[i] ^= (long) asLittleEndian.get(b, ofs);
+            ofs += 8;
+        }
+
+        keccak();
     }
 
     /**
      * Return the digest. Subclasses do not need to reset() themselves,
      * DigestBase calls implReset() when necessary.
      */
     void implDigest(byte[] out, int ofs) {
+        byte[] byteState = new byte[8];
         int numOfPadding =
-            setPaddingBytes(suffix, buffer, (int)(bytesProcessed % buffer.length));
+            setPaddingBytes(suffix, buffer, (int)(bytesProcessed % blockSize));
         if (numOfPadding < 1) {
             throw new ProviderException("Incorrect pad size: " + numOfPadding);
         }
         implCompress(buffer, 0);
-        int availableBytes = buffer.length;
+        int availableBytes = blockSize; // i.e. buffer.length
         int numBytes = engineGetDigestLength();
         while (numBytes > availableBytes) {
-            System.arraycopy(state, 0, out, ofs, availableBytes);
+            for (int i = 0; i < availableBytes / 8 ; i++) {
+                asLittleEndian.set(out, ofs, state[i]);
+                ofs += 8;
+            }
             numBytes -= availableBytes;
-            ofs += availableBytes;
             keccak();
         }
-        System.arraycopy(state, 0, out, ofs, numBytes);
+        int numLongs = (numBytes + 7) / 8;
+
+        for (int i = 0; i < numLongs - 1; i++) {
+            asLittleEndian.set(out, ofs, state[i]);
+            ofs += 8;
+        }
+        if (numBytes == numLongs * 8) {
+            asLittleEndian.set(out, ofs, state[numLongs - 1]);
+        } else {
+            asLittleEndian.set(byteState, 0, state[numLongs - 1]);
```

Comment:
So, is the performance better placing the byteState assignment up there at line 111 instead of here? It's only used in this block...